File: analytic-service/app/models/click_event.py
"""Click event model for MongoDB using Beanie ODM."""

from datetime import datetime
from typing import Optional

from beanie import Document
from pydantic import Field
from pymongo import ASCENDING, DESCENDING, IndexModel


class ClickEvent(Document):
    """
    Click event document model.

    This model represents a click event for link tracking.
    Matches the Go ClickEvent struct for compatibility.
    """

    short_code: str = Field(
        alias="shortCode",
        description="Short code of the link",
    )

    clicked_at: datetime = Field(
        alias="clickedAt",
        description="Timestamp when clicked",
    )

    ip_address_hash: str = Field(
        alias="ipAddressHash",
        description="Hashed IP address",
    )

    user_agent: str = Field(
        alias="userAgent",
        description="User agent string",
    )

    referrer_url: Optional[str] = Field(
        default=None,
        alias="referrerUrl",
        description="Referrer URL",
    )

    referrer_domain: Optional[str] = Field(
        default=None,
        alias="referrerDomain",
        description="Referrer domain",
    )

    country_code: Optional[str] = Field(
        default=None,
        alias="countryCode",
        description="Country code",
    )

    city: Optional[str] = Field(
        default=None,
        alias="city",
        description="City name",
    )

    region: Optional[str] = Field(
        default=None,
        alias="region",
        description="Region/state name",
    )

    device_type: Optional[str] = Field(
        default=None,
        alias="deviceType",
        description="Device type (desktop, mobile, tablet)",
    )

    browser_name: Optional[str] = Field(
        default=None,
        alias="browserName",
        description="Browser name",
    )

    browser_version: Optional[str] = Field(
        default=None,
        alias="browserVersion",
        description="Browser version",
    )

    os_name: Optional[str] = Field(
        default=None,
        alias="osName",
        description="Operating system name",
    )

    os_version: Optional[str] = Field(
        default=None,
        alias="osVersion",
        description="Operating system version",
    )

    lat: Optional[float] = Field(
        default=None,
        alias="lat",
        description="Latitude",
    )

    lon: Optional[float] = Field(
        default=None,
        alias="lon",
        description="Longitude",
    )

    is_bot: bool = Field(
        alias="isBot",
        description="Whether the click is from a bot",
    )

    class Settings:
        """Beanie document settings."""

        name = "click_events"
        use_state_management = True

        # Indexes for better query performance
        indexes = [
            IndexModel([("shortCode", ASCENDING), ("clickedAt", DESCENDING)]),
            IndexModel([("clickedAt", DESCENDING)]),
            IndexModel([("countryCode", ASCENDING)]),
            IndexModel([("isBot", ASCENDING)]),
        ]

    class Config:
        """Pydantic model configuration."""

        populate_by_name = True  # Allow both alias and field name
        json_schema_extra = {
            "example": {
                "shortCode": "my-link",
                "clickedAt": "2025-10-16T10:00:00Z",
                "ipAddressHash": "abc123hash",
                "userAgent": "Mozilla/5.0...",
                "referrerUrl": "https://google.com",
                "referrerDomain": "google.com",
                "countryCode": "ID",
                "city": "Jakarta",
                "region": "Jakarta",
                "deviceType": "desktop",
                "browserName": "Chrome",
                "browserVersion": "120.0",
                "osName": "Linux",
                "osVersion": "x86_64",
                "lat": -6.2088,
                "lon": 106.8456,
                "isBot": False,
            }
        }

File: redirect-service/internal/domain/click_event.go
package domain

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ClickEvent struct {
	ID             primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	ShortCode      string             `bson:"shortCode" json:"shortCode"`
	ClickedAt      time.Time          `bson:"clickedAt" json:"clickedAt"`
	IPAddressHash  string             `bson:"ipAddressHash" json:"ipAddressHash"`
	UserAgent      string             `bson:"userAgent" json:"userAgent"`
	ReferrerURL    string             `bson:"referrerUrl,omitempty" json:"referrerUrl,omitempty"`
	ReferrerDomain string             `bson:"referrerDomain,omitempty" json:"referrerDomain,omitempty"`
	CountryCode    string             `bson:"countryCode,omitempty" json:"countryCode,omitempty"`
	City           string             `bson:"city,omitempty" json:"city,omitempty"`
	Region         string             `bson:"region,omitempty" json:"region,omitempty"`
	DeviceType     string             `bson:"deviceType,omitempty" json:"deviceType,omitempty"` // desktop, mobile, tablet
	BrowserName    string             `bson:"browserName,omitempty" json:"browserName,omitempty"`
	BrowserVersion string             `bson:"browserVersion,omitempty" json:"browserVersion,omitempty"`
	OSName         string             `bson:"osName,omitempty" json:"osName,omitempty"`
	OSVersion      string             `bson:"osVersion,omitempty" json:"osVersion,omitempty"`
	Lat            float64            `bson:"lat,omitempty" json:"lat,omitempty"`
	Lon            float64            `bson:"lon,omitempty" json:"lon,omitempty"`
	IsBot          bool               `bson:"isBot" json:"isBot"`
}

func (ClickEvent) CollectionName() string {
	return "click_events"
}

// PayloadElasticClick represents the structure for sending click event data to Elasticsearch
type PayloadElasticClick struct {
	IndexType string    `json:"index_type"`
	Data      ClickData `json:"data"`
}

type ClickData struct {
	ShortCode string        `json:"short_code"`
	Metadata  ClickMetaData `json:"metadata"`
}

type ClickMetaData struct {
	ClickedAt time.Time     `json:"clicked_at"`
	IsBot     bool          `json:"is_bot"`
	Client    ClientInfo    `json:"client"`
	HTTP      HTTPInfo      `json:"http"`
	UserAgent UserAgentInfo `json:"user_agent"`
}

type ClientInfo struct {
	IPHash string  `json:"ip_hash"`
	Geo    GeoInfo `json:"geo"`
}

type GeoInfo struct {
	CountryISOCode string             `json:"country_iso_code"`
	RegionName     string             `json:"region_name,omitempty"`
	City           string             `json:"city"`
	Location       GeoLocationElastic `json:"location"`
}

type GeoLocationElastic struct {
	Lat float64 `json:"lat"`
	Lon float64 `json:"lon"`
}

type HTTPInfo struct {
	Referrer       string `json:"referrer"`
	ReferrerDomain string `json:"referrer_domain"`
}

type UserAgentInfo struct {
	Original string      `json:"original"`
	Device   DeviceInfo  `json:"device"`
	Browser  BrowserInfo `json:"browser"`
	OS       OSInfo      `json:"os"`
}

type DeviceInfo struct {
	Name string `json:"name"`
}

type BrowserInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

type OSInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

// End PayloadElasticClick
